ICU-11533 Myanmar calendar support - suggestion to add [WIP]

mapmeld opened the pull request:
This is still a work in progress, but as an ICU and C++ newbie, I am reaching out for advice. Any help on setup, please contact me here or @mapmeld on Twitter.

Update: earlier I was overwriting BuddhCal.cpp with MyanCal.cpp to get it to run; for the PR to be more legible I have changed them back to separate files

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-11533
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

CLDR ticket: https://unicode-org.atlassian.net/browse/CLDR-8081
Unicode / ICU ticket: https://unicode-org.atlassian.net/browse/ICU-11533

The code implements the Burmese/Myanmar lunisolar calendar for festival / holiday dates.  Some people in Myanmar use this calendar on official documents and forms, e.g. their birthdate, and it is not simple to convert between them.

### Comparing to a working reference

I am using the open source C++ and JS implementations from https://github.com/yan9a/mcal and making changes:
- renaming variables so their meaning is easier to understand
- changing variable type from long to int where a long is not needed (e.g. day / month / year )
- numbering months in chronological order
- using ClockMath.floorDivide instead of importing math.h ... but is this necessary? Following other calendar implementations
Currently the tests appear to have off-by-one errors (1 month and 1 day off) in most cases.

### Help with development

~~As mentioned earlier, development would be better if I can create calendar=myanmar separate from calendar=buddhist. Also~~ I currently run *all* icu4c tests with ```make check``` instead of the ones which I am targeting - Any help on setup, please contact me here or @mapmeld on Twitter.

### CLA

I have signed the CLA. Because the C++ code is based on Yan Naing Aye's repo, I've filed an Issue asking if he would sign the CLA as well.

mapmeld commented:
I changed the PR so it is more legible - I no longer overwrite BuddhistCalendar

Here are the errors which I see preventing creation of a new calendar type (these errors do not raise when I overwrote an existing calendar)

```
 UObjectTest {
         testIDs {
         FAIL: dynamicID != staticID! MyanmarCalendar * x= Calendar::createInstance(Locale("@calendar=myanmar"), status);  - (Are you missing data?)

         } ERRORS (1) in testIDs (16ms)

...

TestMyanmarFormat {
         Expected Tagu 10, 1350 ME but got April 14, 1989 AD
         Parse incorrect of Tagu 10, 1350 ME - wanted 608603000000 but got 0, December 31, 1969 AD
         U_ILLEGAL_ARGUMENT_ERROR : Error occurred testing Myanmar Calendar in English  - (Are you missing data?)

         } ERRORS (3) in TestMyanmarFormat
```

srl295 commented on .gitignore:
```diff
@@ -1,6 +1,8 @@
 __pycache__/
 *.pyc
+.DS_Store
+icu4c/source/test/*/*test-*.txt
```
These are good but unrelated changes. please move the .gitignore changes to a new PR against one of:

- https://unicode-org.atlassian.net/browse/ICU-20068
- https://unicode-org.atlassian.net/browse/ICU-20527

srl295 commented on icu4c/source/data/lang/en.txt:
```diff
@@ -893,6 +893,7 @@ en{
             islamic-umalqura{"Islamic Calendar (Umm al-Qura)"}
             iso8601{"ISO-8601 Calendar"}
             japanese{"Japanese Calendar"}
+            myanmar{"Myanmar Calendar"}
```
so, this data is from CLDR. So what's here is for the sake of a prototype.

srl295 commented on icu4c/source/data/locales/root.txt:
```diff
@@ -2065,6 +2065,65 @@ root{
             monthNames:alias{"/LOCALE/calendar/gregorian/monthNames"}
             quarters:alias{"/LOCALE/calendar/gregorian/quarters"}
         }
+        myanmar{
```
same comment, this file is derived, so changes need to be on the CLDR side.

srl295 commented on icu4c/source/data/misc/keyTypeData.txt:
```diff
@@ -91,6 +91,7 @@ keyTypeData:table(nofallback){
             islamic-umalqura{""}
             iso8601{""}
             japanese{""}
+            myanmar{""}
```
I think this is derived from CLDR bcp47 data

srl295 commented on icu4c/source/i18n/myancal.h:
```diff
@@ -0,0 +1,327 @@
+// License & terms of use: http://www.unicode.org/copyright.html
+/*
+ ******************************************************************************
+ * Copyright (C) 2003-2013, International Business Machines Corporation
```
Probably don't need this 2003-2013 line and others (I assume this is copy and paste from buddhist as you said)

srl295 commented on icu4c/source/i18n/myancal.h:
```diff
@@ -0,0 +1,327 @@
+
+/**
+ * <code>MyanmarCalendar</code> is a subclass of <code>Calendar</code>
+ * that implements the Persian calendar.  It is used as the official
```
oops… fix references to Persian

srl295 commented on icu4c/source/i18n/myancal.h:
```diff
@@ -0,0 +1,327 @@
+   * Constants for the months
+   * @internal
+   */
+  enum EMonths {
```
fix… persian

srl295 commented on icu4c/source/i18n/myancal.h:
```diff
@@ -0,0 +1,327 @@
+   */
+  virtual ~MyanmarCalendar();
+
+  // TODO: copy c'tor, etc
```
drop this, it has a copy c'tor.

srl295 commented on icu4c/source/i18n/ucal.cpp:
```diff
@@ -259,7 +259,7 @@ ucal_getTimeZoneDisplayName(const     UCalendar*                 cal,
 
 U_CAPI UBool  U_EXPORT2
-ucal_inDaylightTime(    const    UCalendar*      cal, 
+ucal_inDaylightTime(    const    UCalendar*      cal,
```
(i think dropping the whitespace here is fine.)

srl295 commented on icu4c/source/test/testdata/structLocale.txt:
```diff
@@ -21084,6 +21084,7 @@ structLocale:table(nofallback){
             islamic-umalqura{""}
             iso8601{""}
             japanese{""}
+            myanmar{""}
```
Note, this file is not generated.

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+// License & terms of use: http://www.unicode.org/copyright.html
+/*
+ ******************************************************************************
+ * Copyright (C) 2003-2013, International Business Machines Corporation
```
fix date

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+// © 2017 and later: Unicode, Inc. and others.
```
fix date, if there's a © from the original source please put it here

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+ * Modification History:
+ *
+ *   Date        Name        Description
+ *   12/28/2017  mapmeld     copied from persncal.cpp
```
aha, as suspected.

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+    } else { //if (leapStatus == 1) {
+      return 384;
+    }// else if (leapStatus == 2) {
+    //  return 385;
```
?

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+    cal_my(myan_year, myan_year_type, startOfTagu, full_moon_waso_2, addedOffset); //check year
+    dayOfYear = julianDay - startOfTagu + 1;//day count
+    b = ClockMath::floorDivide(myan_year_type, 2);
+    c = ClockMath::floorDivide(1, (myan_year_type + 1)); //big wa and common yr
```
may be better `int32_t c = …` and don't declare above.

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+    } else if (myan_month >= 4) {
+      myan_month++;
+    }
+    //myan_month--;
```
?

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+      myan_month++;
+    }
+    //myan_month--;
+    //myan_day++;
```
?

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+	watat^=EW;//correct watat exceptions
+}
+
+// Myanmar does not change clocks for Daylight Time
```
hm. this may be true, but it is a property of the TZ not of the calendar.

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+    MyanmarCalendar calendar(Locale("@calendar=myanmar"),status);
+   if (U_SUCCESS(status))
+   {
+      printf("success\n");
```
printf… put it  behind `#if defined (U_DEBUG_CAL)  … #endif` and mention the file and line, see https://github.com/unicode-org/icu/blob/85c7e61b3e788be448a4ea96eb8861c1a4003d94/icu4c/source/i18n/gregocal.cpp#L377

srl295 replied:
(will also need `#if defined (U_DEBUG_CAL)  …   #include <stdio.h> … #endif` earlier in the source file)

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+        gSystemDefaultCenturyStart = calendar.getTime(status);
+        gSystemDefaultCenturyStartYear = calendar.get(UCAL_YEAR, status);
+   } else {
+     printf("fail create Myanmar\n");
```
see above, debug printf

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,501 @@
+     printf("fail create Myanmar\n");
+   }
+    // We have no recourse upon failure unless we want to propagate the failure
+    // out.
```
right. It's probably a low memory condition.

srl295 commented:
@mapmeld thanks for this!  The automated builds are failing (look for the red X's above), see https://travis-ci.org/unicode-org/icu/jobs/522911134#L1351 for example.

mapmeld commented:
@srl295 progress update / question: am I allowed to import math.h here? The other calendars use ```ClockMath::floorDivide``` instead, but it is giving me an unusual (negative) result on one line dividing doubles, causing everything to be shifted one lunar month later on in the code

```
ClockMath::floorDivide(double(SOLAR_YEAR * (long(myan_year) + 3739)), LUNAR_MONTH, excess_days);
```

srl295 commented:
@mapmeld `math.h` is probably OK, as `astro.cpp` and others use it.  If you can use the  `uprv_` functions such as `uprv_floor()` that gives a little more platform specific separation. 

Note that `floorDivide` is implemented via `uprv_floor(a/b)` which just calls `floor`.

srl295 commented on icu4c/source/i18n/myancal.cpp:
```diff
@@ -0,0 +1,506 @@
+  long l=0;
+  u--;
+	while (u >= l) {
+		i = ClockMath::floorDivide(l + u, 2);
```
change `long` to `int32_t` here and similar sites.

srl295 commented on icu4c/source/i18n/myancal.h:
```diff
@@ -0,0 +1,327 @@
+ *
+ * @see GregorianCalendar
+ *
+ * @author Mehran Mehr
```
fix

srl295 commented:
@mapmeld ping on this?

mapmeld commented:
I got formal approval from the original author of the C++ code. Unfortunately I haven't looked at this for a while, and never found a good way to debug what was happening. I feel like this is VERY close but only for someone who understands why it is usually off by one lunar month.

srl295 commented:
> I got formal approval from the original author of the C++ code. Unfortunately I haven't looked at this for a while, and never found a good way to debug what was happening. I feel like this is VERY close but only for someone who understands C++ and can find out why it is usually off by one lunar month.

didn't get to debug it but the 'month' field does start at 0 of course, that always gets me…  Can you rebase it against current master?

Also, if you didn't already do so, we'll definitely need a CLDR ticket for this, and the initial english/etc names should go there. IF you start a PR that adds the right stuff into CLDR's en.xml, etc, i can help with the data generation

mapmeld commented:
@srl295 sorry for another long delay on responding to the project. I've updated to the current master branch and reapplied my changes in one commit.  I have tests for MyanCal but they have been failing, off by one lunar month.  I don't know how to test my code without the entire ICU test suite, which makes it difficult for me to switch back and forth between coding and testing or printf-ing to figure out where I'm losing the correct values.

markusicu commented:
> @srl295 sorry for another long delay on responding to the project. I've updated to the current master branch and reapplied my changes in one commit. I have tests for MyanCal but they have been failing, off by one lunar month. I don't know how to test my code without the entire ICU test suite, which makes it difficult for me to switch back and forth between coding and testing or printf-ing to figure out where I'm losing the correct values.

This is what I do on Linux:
- git repo workspace in ~/icu/mine/src
- out of source build into ~/icu/mine/dbg/icu4c
- `cd ~/icu/mine/dbg/icu4c/test/intltest`
- `export LD_LIBRARY_PATH=../../lib:../../stubdata:../../tools/ctestfw`
- `make && ./intltest format/IntlCalendarTest`

mapmeld commented:
I rebased to make this current, in hopes that someone (possibly me, but happy to pass this on) can debug the tests

mapmeld commented:
After a long break, I've revisited this PR, rebased, added tests, and addressed most of the issues.
To keep dates verifiable I will probably limit the PR to more recent eras (1800 AD and onward)

Remaining issues:
- There's an off-by-one bug between summer 2016-2018 (different types of leap years). The current patch should be replaced by fixing the source of the off-by-one error
- a note says month names Late Tagu and Late Kason are for old eras, but they appear here and in the JS app (short explanation: if the date occurs at the end of the year it uses the name Late Tagu; for example incrementing the date from Late Tagu 11, 1350 ME -> Tagu 12, 1351 ME)
- I should add tests prior to 1989 AD, especially around the change of eras in 1895 and 1951. There is a test failing for 1609 AD which I don't remember where I found
- It's difficult to find reference for when a great leap year (additional day on Nayon) last occurred prior to 2015 AD

macchiati reviewed (changes requested):
> I got formal approval from the original author of the C++ code. Unfortunately I haven't looked at this for a while, and never found a good way to debug what was happening. I feel like this is VERY close but only for someone who understands C++ and can find out why it is usually off by one lunar month.

If you are not the original author, then we will need the original author to sign a CLA also.

mapmeld commented:
@macchiati I didn't update the list and forgot this, but Yan9a signed the CLA in 2020: https://github.com/yan9a/mmcal/issues/2#issuecomment-574132201

srl295 commented:
closed?

mapmeld commented:
I've made a comment on the issue tracker about implementing a Myanmar year + era calendar (similar to the Thai / buddhist / buddhcal currently in ICU) https://unicode-org.atlassian.net/browse/ICU-11533?focusedCommentId=182810 as that might be easier to implement and test
